lib/l1_watcher/src/batch_range_watcher.rs

```diff
@@ -1,25 +1,29 @@
 use crate::watcher::{L1Watcher, L1WatcherError};
 use crate::{L1WatcherConfig, ProcessL1Event, util};
 use alloy::consensus::Transaction;
-use alloy::primitives::{Address, BlockNumber};
+use alloy::eips::BlockId;
+use alloy::primitives::{Address, B256, BlockNumber};
 use alloy::providers::{DynProvider, Provider};
 use alloy::rpc::types::Log;
 use alloy::sol_types::{SolCall, SolValue};
+use anyhow::Context;
 use std::sync::Arc;
+use tokio::sync::mpsc;
 use zksync_os_contract_interface::IExecutor::ReportCommittedBatchRangeZKsyncOS;
-use zksync_os_contract_interface::models::{CommitBatchInfo, StoredBatchInfo};
+use zksync_os_contract_interface::models::CommitBatchInfo;
 use zksync_os_contract_interface::{IExecutor, ZkChain};
+use zksync_os_types::ProtocolSemanticVersion;
 
 /// Don't try to process that many block linearly
 const MAX_L1_BLOCKS_LOOKBEHIND: u64 = 100_000;
 
-/// Discovers block ranges for batches `[last_executed_batch + 1; last_committed_batch]`. This is
+/// Discovers commitment data for batches `[last_executed_batch; last_committed_batch]`. This is
 /// needed to rebuild batches correctly in Batcher during replay.
 pub struct BatchRangeWatcher {
-    contract_address: Address,
-    provider: DynProvider,
+    zk_chain: ZkChain<DynProvider>,
     next_batch_number: u64,
     last_batch_number: u64,
```

Comment:
maybe `last_committed_batch_on_startup`?  Please add detailed comment for both next/last_batch_number - now it's not obvious what they are and how/if/when they change.